### DISCUSSION TITLE
OnClick event for NavBar items

### DISCUSSION THREAD

**Discussion by matkoch** [facebook/docusaurus#6642] [Q&A] [answered]
The [documentation](https://docusaurus.io/docs/api/themes/configuration#navbar-items) says:

> In addition to the fields above, you can specify other arbitrary attributes that can be applied to a HTML link.

I tried adding an `onClick` handler like so:

```
onClick: "fathom.trackGoal('B5PGZD9N', 0);"
```

But this doesn't seem to work :/ Do I have to swizzle the component or am I just doing something wrong?

**Comment by Josh-Cena** [comment, marked as answer]
See https://github.com/facebook/docusaurus/discussions/6527

Yes, callbacks can't be passed from the config file to client-side. You need to swizzle the component.

**Reply by matkoch** [reply]
I think I accepted this prematurely. Which component exactly should I swizzle? I noticed that also with the `Link` component, the `onclick` attribute won't propagate. Is this connected to the issue?

Edit: oh actually @slorber accepted it 😄

**Reply by slorber** [reply]
you have to swizzle the navbar item component you are using. (you didn't tell us which one so we can't tell more)

I suggest wrapping one of those 2 items and intercepting the click listeners according to your needs: NavbarNavLink or DefaultNavbarItem

**Reply by slorber** [reply]
@matkoch what is your real purpose here? it's difficult to help you otherwise as we don't really understand what you are trying to do.

What happens when the navbar item is clicked?

Can't you track the goal once the user visits a given page? 

For example with a React `useEffect` in that page, or using a client module?

```tsx
import type {ClientModule} from '@docusaurus/types';

const clientModule: ClientModule = {
  onRouteDidUpdate({location}) {
    if (location.pathname === "/target-page" ) {
      fathom.trackGoal('B5PGZD9N', 0);
    }
  },
};

export default clientModule;
```

Can't Fathom create somehow "artificial events" like many other analytics systems? 
IE you can tell it that `visit /my-page` means "my-goal".

**Reply by matkoch** [reply]
Fathom already does track page navigation (solved in #6671).

Goals can be used for something different, for instance when you want to track links going off the page. Particularly, I'm using them to track clicks on an install button handled by a custom protocol handler (`local-app://xxxx`).

**Reply by slorber** [reply]
I see thanks. 

We have no easy way to achieve that afaik. Eventually we could build an event system so that all link clicks lead to a client module "onLinkClick" event or something 🤷‍♂️. Or maybe we should always wrap the core link in a theme link so that it could be swizzled? 🤷‍♂️ 

One temporary workaround could be to create a static HTML file that just calls your analytics, and then redirect to the custom app scheme protocol. That's not ideal for sure.

**Reply by matkoch** [reply]
Yea, that could be a solution, thanks for getting back so quickly! :)

**Reply by tisonkun** [reply]
@slorber another possiblity is support "button" type on navbar item.

I have such a page:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/18818196/208846501-916cb4dd-6ed0-40f2-b2b9-41f74a60b6e0.png">

Now when I click the log out item, I have to jump to a logout page and then click a button there to do logout:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/18818196/208846975-4be2c7dd-97cf-4870-828a-83ddfff24d90.png">

Code:

```ts
// in docusaurus.config.js
                    {
                        to: '/logout',
                        position: 'right',
                        className: 'header-logout-link',
                        'aria-label': 'Logout page',
                    },

// in logout.tsx

        <p className={styles.description}>
            <Button variant={'contained'} onClick={() => logout({
                returnTo: logoutUri,
            })}>
                点击登出
            </Button>
        </p>
```

Where I'd prefer to trigger `onClick` just when I click the navbar item.

**Reply by slorber** [reply]
Not possible to provide event listeners from nodejs config file.

We could eventually decide to create another client-side theme config, like `./themeConfig.js` in which you could be able to provide callbacks/JSX/React code, but we don't have that for now

---

In this case I would suggest to create a custom navbar item with the following workaround:
https://github.com/facebook/docusaurus/issues/7227

```
// in docusaurus.config.js
{
  type: "custom-logout-button"
},
```

**Reply by tisonkun** [reply]
@slorber Thanks for your shares!

> In this case I would suggest to create a custom navbar item with the following workaround:

It seems this feature hasn't been delivered now?

UPDATE: I found https://github.com/facebook/docusaurus/pull/7231 now :)

**Reply by tisonkun** [reply]
Implemented and works well on Desktop. I'm digging whether it can display properly on mobile: it seems getting hidden and since I use a docs only site I cannot find it around the page on mobile.

UPDATE: it can be found by jumping back the main menu.

FYI, the code adjusted to mobile style is:

```ts
    if (props.mobile) {
        return <li className="menu__list-item">
            <NavbarNavLink
                onClick={() => logout({
                    returnTo: logoutUri,
                })}
                className={clsx('menu__link', 'header-logout-link')}
                aria-label={'Logout page'}
            />
        </li>
    }

    return <NavbarNavLink
        onClick={() => logout({
            returnTo: logoutUri,
        })}
        className={clsx('navbar__item navbar__link', 'header-logout-link')}
        aria-label={'Logout page'}
    />;
```

`NavbarNavLink` instead `button` provide the proper render effect.

**Reply by sneko** [reply]
Thanks for pointing me in the right direction. Just note as written in the PR you MUST prefix the new component type with `custom-` otherwise it will fail 😄
